### PR TITLE
[GUI] Radial gradient implementation

### LINKED ISSUE #8297: [Feature Request] Radial/Elliptical gradient as display background colour
### Is there an existing request for this?

- [X] I have searched the existing requests

### Forums discussion

https://forum.freecadweb.org/viewtopic.php?t=72672&start=60

### Subproject(s) affected?

Other (specify in description)

### Idea description

Currently, it is possible to choose between a flat background colour and a linear gradient:
Linear gradient

It would be nice to have a radial/ elliptical gradient as option:
elliptical gradient

The reason is that this gives a more balanced colour distribution across the screen. That way, there is less of a strong highlight in the centre and more a halo-like effect, which helps improve visibility of the model and sketches.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by xtemp09**
This commit implements radial gradient as background and adds the option to settings. It also renames "Color gradient" as "Linear gradient", keeping Linear gradient default. Internally, it remains unchanged for compatibility.

Radio gradient is more suitable for CAD, since it gives a more balanced color distribution across the screen, improving visibility of model and sketches with a halo-like effect.

This pull request closes #8297. I advise to set radial gradient default in the upcoming version of FreeCAD.

<details><summary>Options</summary>

Screenshot 1 - Options

</details>

<details><summary>Screenshot 2</summary>

Screenshot 2

</details>

<details><summary>Screenshot 3</summary>

Screenshot 3

</details>

<details><summary>Screenshot 4</summary>

Screenshot 4

</details>

<details><summary>ProDark</summary>

ProDark

</details>

<details><summary>ProDark2</summary>

ProDark 2

</details>

<details><summary>ProDark3</summary>

ProDark 3

</details>

The commit looks huge due to git, though it just adds several lines of code, mostly to `SoFCBackgroundGradient.cpp` and an option to the settings.

<details><summary>Standard form</summary>

Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

</details>

### CODE REVIEW

**Comment by luzpaz** [comment]
CC @Intuos5 @MisterMakerNL (participants in #8297)

**Comment by Intuos5** [comment]
Thanks @xtemp09 looks really nice! And thanks @luzpaz for tagging me!

**Comment by MisterMakerNL** [comment]
Looks awesome!
I can compile and try. on windows when I'm back home!

**Comment by chennes** [comment]
What happened to View3DViewer.py, that this PR touches every line?

**Comment by adrianinsaval** [comment]
line ending auto changed by the editor maybe?

**Comment by chennes** [comment]
Could be... @xtemp09 could you please do a `git checkout` on that file and re-commit just the single-line change you made, without the line ending modification? Looks like it's just
```
_viewer->setGradientBackgroundColor(col, col, false);
```

**Comment by xtemp09** [comment]
@chennes, I've just redone those files, but I don't know what happens if I push. Documentation on git is quite confusing.

Can I change the commit online with github.dev? If not, I think, the reviewers should just know that an only line has been changed.

Before I thought it's Qt Creator fault, now I think that git's `core.autocrlf` does not work properly, because I didn't change line endings in this commit.

**Comment by MisterMakerNL** [comment]
I tested it on windows and works great although I would recommend some clarifications in the settings. something like this:
afbeelding
afbeelding
afbeelding

**Comment by xtemp09** [comment]
@MisterMakerNL, are those errors with package.xml coming from my commit?

**Comment by 0penBrain** [comment]
> What happened to View3DViewer.py, that this PR touches every line?

Just for information when looking 'Files changed' in GH, there are parameters at the top of the page with a 'Hide whitespaces' option.

**Comment by chennes** [comment]
> Just for information when looking 'Files changed' in GH, there are parameters at the top of the page with a 'Hide whitespaces' option.

Yeah, that's how I found the real line changed. Looking forward to being done with unintended reformats [once and for all](https://github.com/FreeCAD/FreeCAD/pull/9140)!

**Comment by chennes** [comment]
This works well -- I agree with @MisterMakerNL that the interface can now use some help, but I think that can happen in a follow-on PR (maybe @PaddleStroke would like to give it a look?). I'm going to merge as-is so we can move forward.

**Comment by PaddleStroke** [comment]
Yes I agree the ui can be improved. MisterMaker proposal is a step in the good direction. I will think about it and make a PR later.

**Comment by xtemp09** [comment]
@PaddleStroke, if you are going to fix settings, one point to remember. I forcedly git pushed and forgot one little thingy.
https://github.com/FreeCAD/FreeCAD/blob/b75d257e80e4a2d23c780a6da4002c7e75d78e9b/src/Gui/SplitView3DInventor.cpp#L267-L270

In this case gradient background is set twice. First, when _Gradient_ option is read, then when _RadialGradient_ is read. If you are going to change `src/Gui/DlgSettingsViewColor.ui`, could you also change this? Setting gradient background twice, basicly, does not do any harm, perhaps it's just illogical.

I was about to patch `src/Gui/SplitView3DInventor.cpp`, but read that you volonteered. 😇 Also, I don't really imagine how to clarify `src/Gui/DlgSettingsViewColor.ui`. For linear gradient the first, the second and the third colors are _Top color_,  _Bottom color_ and _Middle color_, while for radial gradient they are _Central color_, _End color_ and _Stop point color_, at least, in [Qt terminology](https://doc.qt.io/qt-6/qradialgradient.html).

**Comment by PaddleStroke** [comment]
@xtemp09 Can you please specify the change required exactly?

**Comment by xtemp09** [comment]
Sorry, my brain processed incorrect values producing incorrect results. =)

**Comment by PaddleStroke** [comment]
I just checked on master and it works correctly.

**Comment by wwmayer** [comment]
> @wwmayer, are you sure it's caused by my commit? I checked, it worked.

It's the only recent change made to background colors of the 3d view.

**Comment by xtemp09** [comment]
wwmayer, how can I reproduce the problem?

**Comment by wwmayer** [comment]
I am currently investigating the issue...

File: src/Gui/View3DSettings.cpp [hunk 1]
```diff
@@ -233,8 +234,8 @@ void View3DSettings::OnChange(ParameterGrp::SubjectType &rCaller,ParameterGrp::M
     else if (strcmp(Reason,"UseAutoRotation") == 0) {
         _viewer->setAnimationEnabled(rGrp.GetBool("UseAutoRotation", false));
     }
-    else if (strcmp(Reason,"Gradient") == 0) {
-        _viewer->setGradientBackground((rGrp.GetBool("Gradient", true)));
+    else if (strcmp(Reason,"Gradient") == 0 || strcmp(Reason,"RadialGradient") == 0) {
+        _viewer->setGradientBackground(rGrp.GetBool("Gradient", true) || rGrp.GetBool("RadialGradient", true));
```
Comment (wwmayer) [inline src/Gui/View3DSettings.cpp:238]: Using true as the default value for RadialGradient causes the regression. On a system where this parameter does not yet exist the returned value of GetBool() is true and thus activates the gradient even if it's disabled in the settings.

Comment (xtemp09) [reply src/Gui/View3DSettings.cpp:238]: In other places I wrote `rGrp.GetBool("RadialGradient", false)`. So, it's just a typo.

I'll post a fix for this.

**Comment by wwmayer** [comment]
https://github.com/FreeCAD/FreeCAD/pull/9178

File: src/Gui/View3DSettings.cpp (hunk 1 above)
Comment (wwmayer) [reply src/Gui/View3DSettings.cpp:238]: I have it already.

Comment (0penBrain) [reply src/Gui/View3DSettings.cpp:238]: > Using true as the default value for RadialGradient causes the regression. On a system where this parameter does not yet exist the returned value of GetBool() is true and thus activates the gradient even if it's disabled in the settings.

Would that be a big deal to add a command switch like '--vanilla' that would run FC without loading/using config files (and that won't overwrite user config)?

Comment (wwmayer) [reply src/Gui/View3DSettings.cpp:238]: You can use the option "-u" that allows you to use an arbitrary config file.

Comment (0penBrain) [reply src/Gui/View3DSettings.cpp:238]: > You can use the option "-u" that allows you to use an arbitrary config file.

Yes I know. :smile: I do launch with `-u /dev/null -s /dev/null`. Works but launch with this error :wink: 
```
Fatal Error at file "/dev/null", line 1, column 1
   Message: invalid document structure
Malformed Parameter document: Root group not found
Malformed Parameter document: Root group not found in file /dev/null.
Continue with an empty configuration.
Fatal Error at file "/dev/null", line 1, column 1
   Message: invalid document structure
Malformed Parameter document: Root group not found
Malformed Parameter document: Root group not found in file /dev/null.
Continue with an empty configuration.
```

**Pull request merged**